### ISSUE TITLE
Consider what a LocalMachine\My X509Store might be on Unix

### ISSUE THREAD

**Issue by bartonjs** [dotnet/corefx#3690]
Largely there's a permissions problem.

* File writes, without reaching into the PAL directly, are done as rwx------
* Windows has the flexible administrator concept, Unix has root.
    * There's sometimes the group wheel, but that's not universal.

There's a possibility of having someone create /etc/dotnet/corefx/cryptography/x509stores, setting the group to whatever they want, giving the group write permission, and then we could start making use of that... but it's not as obvious as how to do it on Windows.

**Comment by stephentoub** [comment]
> File writes, without reaching into the PAL directly, are done as rwx

Do you think the default permissions for new files created by System.IO.FileSystem should be something else?

**Comment by bartonjs** [comment]
@stephentoub The default default should assume user data, so I agree with it.  But for a machine-scoped thing it would likely need to be rw?rw?r-?; and chown'd to root.  I don't think those concepts exist in the public surface area.

If that was the only thing standing in the way it would really be an issue, that's what PALs are for :).  The initial directory creation is one big problem.  (Usually) only root has access to create directories within /etc.  Once /etc/dotnet/corefx/cryptography/x509stores was created and set to the right permissions (rwxrwxr-x, uid:root gid:<group to manage certificates>) the rest could be maintained by the X509Store PAL.

I've now remembered to add the hard bullet... Windows allows everyone to enumerate certificates in the LM\My store... that is, to see the public key/certificate.  The approach taken for user stores of using a PFX for either public-only or public-private wouldn't work.

**Comment by natemcmaster** [comment]
Has there been any more discussion or progess on designing this feature? I hit this limitation while prototyping support for certificates + encrypted XML with ASP.NET Core's data protection libraries.

See also: https://github.com/dotnet/corefx/issues/19489

**Comment by bartonjs** [comment]
@natemcmaster Nope, it remains a difficult to reason-out problem; because it's really an OS feature of Windows (and macOS).  But, as I commented in your other issue, the problem isn't in X509Store per-se, but in EncryptedXml not handling the failure as gracefully as it handles other failures.

**Comment by the-black-wolf** [comment]
has this been given more though? Its kind of annoying, it forces us into writing if/then/else spaghetti code for our .net standard library, which is a complete antithesis of what .net standard is about. Whats the point of having an X509 class in .net **standard** if it throws below exceptions? Then its not a cross platform standard.
_**System.PlatformNotSupportedException:** Unix LocalMachine X509Store is limited to the Root and CertificateAuthority stores._

using /etc/dotnet/corefx/cryptography/x509stores seems like a reasonable solution which pushes the issue where it should be, with administration.

**Comment by bartonjs** [comment]
@the-black-wolf On Windows any user can enumerate LM\My and use the X509Certificate2 objects representing the certificates; cert.HasPrivateKey will be true, but accessing the private key will result in Access Denied (unless the user has permissions, of course).

On Linux the user cert stores are implemented as one-PFX per cert (which may or may not contain a private key).  We can't use that to get the same behavior for a LM\My store, since "see cert" and "see key" are the same... so let's say we split it into the private keys are stored in a subdirectory of the store directory. We don't have any support on Linux (so it'd have to be built) for knowing that there "is" a private key, but you can't access it.  Our cert model says HasPrivateKey is true if we have a private key loaded, false otherwise.

Okay, so we can make the directory and subdirectory and use restrictive permissions on add (limit it to when geteuid is 0, make private keys rw-------), make another kind of store loader that can load that kind of store, teach the system about "there's a private key you can't have" (because maybe that's what the program is trying to determine)... and what does it really gain?  On Windows the certificate stores are applicable to the whole OS, all programs can use it equally.  On Linux it'd be just .NET programs (or anyone who wanted to participate in our format, I guess).

But now we get to: what path? Maybe CentOS wants us in `/var/...`, and Debian wants us under `/etc/contrib/...`, while Ubuntu is happy with `/etc/...`.

So there's a bit of policy involved, and this feels more like "shoving .NET at Linux" instead of "embracing Linux from .NET"... we've mainly been striving for "embrace".

> it forces us into writing if/then/else spaghetti code for our .net standard library

There are several different possible approaches.  If you're trying multiple stores with fallback, just put the call to X509Store.Open in a try/catch-PNSE.  The Certificates collection will come back as the empty collection (because that's what we do for closed stores, for better or worse), you won't find a match, you try the next store.  If you're trying just one, then making CurrentUser vs LocalMachine a variable based on RuntimeInformation.IsOSPlatform is less than ideal, but not overly cumbersome.

```C#
using (X509Store lmMy = new X509Store(StoreName.My, StoreLocation.CurrentMachine))
{
    try
    {
        lmMy.Open(OpenFlags.ReadOnly);
    }
    catch (PlatformNotSupportedException)
    {
    }

    // At this point, there's not an observable difference between "the store is open and empty" and "the store is closed"
}
```

---

All that said, what's your scenario?  Why would you even want a machine-scoped My store on Linux?

**Comment by the-black-wolf** [comment]
@bartonjs  

Well, you can safely move the onus for this on administration. subfolder structure sounds just fine to me, make it mandatory for administration to export cer from pfx and place two files, or provide a tool in dotnet to do it for the user. That way you get browse and load private capability separate. 

Embracing wont work here, exactly because there is no standard to embrace. Even `openssl storeutl` asks for a URI. You can't wait for the moment that Linux adopts this to embrace it because that might and probably will never come. So on the question of "where", use subfolder in dotnet? FF/Chrome/etc use **NSS Shared DB**, that's another option thats getting more and more traction. 
Even without it, reducing the scope of access from "all applications" to ".net core applications" sounds just fine to me, whatever location or mechanism you choose others wont be able to use automatically in any scenario because there is no system level API to abstract it for them (unless we talk about NSS).  

The scenario is a multi service machine, with multiple IdP and crypto services around the same certificate w/key. Certs expire and we need a simple way of updating all services on server to use the new one, so machine store is a perfect place (which is why MS created it in windows in the first place). We also need this package of services to work on both Linux and Windows. And lets be honest, exception fallback is just a fancy name for a slower version of if/then/else ;)
